backends/p4fmt/p4fmt.cpp

```diff
@@ -1,15 +1,53 @@
 #include "backends/p4fmt/p4fmt.h"
 
+#include "backends/p4fmt/attach.h"
+#include "backends/p4fmt/p4formatter.h"
 #include "frontends/common/parseInput.h"
 #include "frontends/common/parser_options.h"
+#include "frontends/parsers/parserDriver.h"
 #include "ir/ir.h"
 #include "lib/compile_context.h"
 #include "lib/error.h"
+#include "lib/source_file.h"
 #include "options.h"
-#include "p4formatter.h"
 
 namespace P4::P4Fmt {
 
+std::pair<const IR::P4Program *, const Util::InputSources *> parseProgram(
```

Comment:
This code is copied but we can probably simplify it for the P4Fmt context. E.g., you do not need the options check and I wonder whether we should have the preprocessor enabled here. Probably, because we do not support macros yet. 

Also, instead of returning nullptr for the pairs, use std::optional and return a std::nullopt.